core_java/twitter/src/main/java/ca/jrvs/apps/twitter/dao/TwitterDao.java

```diff
@@ -0,0 +1,166 @@
+package ca.jrvs.apps.twitter.dao;
+
+import ca.jrvs.apps.twitter.dao.helper.HttpHelper;
+import ca.jrvs.apps.twitter.model.Coordinates;
+import ca.jrvs.apps.twitter.model.Tweet;
+import ca.jrvs.apps.twitter.util.JsonParser;
+import com.google.gdata.util.common.base.PercentEscaper;
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import org.apache.http.HttpResponse;
+import org.apache.http.util.EntityUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public class TwitterDao implements CrdDao<Tweet, String> {
+
+  // TWITTER resource urls
+  private static final String TWITTER_BASE_URI = "https://api.twitter.com/1.1/";
+  private static final String TWITTER_POST_URI = TWITTER_BASE_URI + "statuses/update.json";
+  private static final String TWITTER_SHOW_URI = TWITTER_BASE_URI + "statuses/show.json";
+  private static final String TWITTER_DELETE_URI = TWITTER_BASE_URI + "statuses/destroy/";
+
+  // TWITTER resource symbols
+  private static final String QUERY_SYM = "?";
+  private static final String AMPERSAND = "&";
+  private static final String EQUAL = "=";
+
+  // Response code
+  private static final int HTTP_OK = 200;
+
+  private HttpHelper httpHelper;
+  private Logger logger = LoggerFactory.getLogger(TwitterDao.class);
+
+  @Autowired
+  public TwitterDao(HttpHelper httpHelper) {
+    this.httpHelper = httpHelper;
+  }
+
+  public Tweet parseResponseBody(HttpResponse response, Integer expectedStatusCode) {
+    Tweet tweet = null;
+
+    // checks response status
+    int status = response.getStatusLine().getStatusCode();
+    if (status != expectedStatusCode) {
+      try {
+        System.out.println(EntityUtils.toString(response.getEntity()));
+      } catch (IOException e) {
+        logger.error("Response has no entity");
+      }
+      throw new RuntimeException("Unexpected HTTP status: " + status);
+    }
+    if (response.getEntity() == null) {
+      throw new RuntimeException("Empty response body");
+    }
+
+    // converts response entity to string
+    String responseString;
+    try {
+      responseString = EntityUtils.toString(response.getEntity());
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to convert JSON to string");
+    }
+
+    // deserializes the string format of the response to a Tweet object
+    try {
+      tweet = JsonParser.toObjectFromJson(responseString, Tweet.class);
+    } catch (IOException e) {
+      throw new RuntimeException("Unable to convert JSON string to a Tweet object");
```

Comment:
you need pass the caught exception when throwing a new one. otherwise, you will lose the information 
      throw new RuntimeException("Unable to convert JSON string to a Tweet object", e);  //see the second arg